File: 1Lab/AVLTree.cpp
#include "AVLTree.h"
#include <iostream>
#include <algorithm>

using namespace std;

int GETH(AVLNode* node) {
    return node ? node -> height : 0;
}

// вычислить бф: разница высот левого и правого
int GETBF(AVLNode* node) {
    if (!node) return 0;
    return GETH(node -> left) - GETH(node -> right);
}

// обновить высоту узла после изменений
void updateHeight(AVLNode* node) {
    if (node) {
        node -> height = 1 + max(GETH(node -> left), GETH(node -> right));
    }
}

AVLNode* rotateLeft(AVLNode* x) {
    AVLNode* y = x -> right;
    AVLNode* T2 = y -> left;

    y -> left = x;
    x -> right = T2;

    updateHeight(x);
    updateHeight(y);

    return y; // y становится новым корнем
}

AVLNode* rotateRight(AVLNode* y) {
    AVLNode* x = y -> left;
    AVLNode* T2 = x -> right;

    x -> right = y;
    y -> left = T2;

    updateHeight(y);
    updateHeight(x);
    
    return x; // x становится новым корнем
}

AVLNode* TINSERT(AVLNode* node, const string& key) {
    if (!node) {
        return new AVLNode(key);
    } 
    if (key < node -> key) {
        node  -> left = TINSERT(node -> left, key);
    } else if (key > node->key) {
        node -> right = TINSERT(node -> right, key);
    } else {
        return node; // дубликаты не допускаются
    }

    updateHeight(node);

    int balance = GETBF(node);
    
    // Левый-Левый случай
    if (balance > 1 && key < node -> left -> key) {
        return rotateRight(node);
    }

    // Правый-Правый случай
    if (balance < -1 && key > node -> right -> key) {
        return rotateLeft(node);
    }

    // Левый-Правый случай
    if (balance > 1 && key > node -> left -> key) {
        node -> left = rotateLeft(node -> left);
        return rotateRight(node);
    }

    // Правый-Левый случай
    if (balance < -1 && key < node -> right -> key) {
        node -> right = rotateRight(node -> right);
        return rotateLeft(node);
    }
    return node;
}

AVLNode* TGET(AVLNode* root, const string& key) {
    if (root == nullptr || root -> key == key) {
        return root;
    }
    if (key < root->key) {
        return TGET(root -> left, key);
    }
    return TGET(root -> right, key);
}

void TPRINT(AVLNode* root, string indent, bool last) {
    if (root != nullptr) {
        cout << indent;
        if (last) {
            cout << "R----";
            indent += "   ";
        } else {
            cout << "L----";
            indent += "|  ";
        }
        cout << "\"" << root->key << "\"" << "(h:" << root->height << ", bf:" << GETBF(root) << ")" << endl;
        TPRINT(root -> left, indent, false);
        TPRINT(root -> right, indent, true);
    }
}

// поиск узла с минимальным значением в поддереве
AVLNode* findMinValueNode(AVLNode* node) {
    AVLNode* current = node;
    while (current && current->left != nullptr) {
        current = current->left;
    }
    return current;
}

AVLNode* TDEL(AVLNode* root, const string& key) {
    if (root == nullptr) {
        return root;
    }

    if (key < root -> key) {
        root -> left = TDEL(root -> left, key);
    } else if (key > root->key) {
        root -> right = TDEL(root -> right, key);
    } else {
        // У узла нет или есть один дочерний узел
        if (root -> left == nullptr || root->right == nullptr) {
            AVLNode* temp = root -> left ? root -> left : root -> right;

            // Если нет дочерних узлов
            if (temp == nullptr) {
                temp = root;
                root = nullptr;
            } else { // если 1 дочерний узел
                *root = *temp; // копируем содржимое доч узла
            }
            delete temp;
        } else { // у узла два дочерних узла
            // Находим наименьший узел в правом поддереве
            AVLNode* temp = findMinValueNode(root -> right);
            root -> key = temp -> key; // Копируем его ключ в текущий узел 
            root -> right = TDEL(root -> right, temp -> key);
        }
    }
    
    if (root == nullptr) {
        return root;
    }

    // Обновление высоты и балансировка
    updateHeight(root);
    int balance = GETBF(root);

    // Левый-Левый случай
    if (balance > 1 && GETBF(root -> left) >= 0) {
        return rotateRight(root);
    }

    // Левый-Правый случай
    if (balance > 1 && GETBF(root -> left) < 0) {
        root->left = rotateLeft(root -> left);
        return rotateRight(root);
    }

    // Правый-Правый случай
    if (balance < -1 && GETBF(root -> right) <= 0) {
        return rotateLeft(root);
    }

    // Правый-Левый случай
    if (balance < -1 && GETBF(root -> right) > 0) {
        root->right = rotateRight(root -> right);
        return rotateLeft(root);
    }

    return root;
} 

File: 1LabGo/avl_tree.go
package main

import "fmt"

type AVLNode struct {
	key    string
	height int
	left   *AVLNode
	right  *AVLNode
}

type AVLTree struct {
	root *AVLNode
}

func height(n *AVLNode) int {
	if n == nil {
		return 0
	}
	return n.height
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func (t *AVLTree) TInsert(key string) {
	t.root = tInsert(t.root, key)
}

func (t *AVLTree) TDel(key string) {
	t.root = tDel(t.root, key)
}

func (t *AVLTree) TGet(key string) bool {
	return tGet(t.root, key) != nil
}

func (t *AVLTree) TPrint() {
	tPrint(t.root, "", true)
}

// Рекурсивные вспомогательные функции
func tInsert(node *AVLNode, key string) *AVLNode {
	if node == nil {
		return &AVLNode{key: key, height: 1}
	}
	if key < node.key {
		node.left = tInsert(node.left, key)
	} else if key > node.key {
		node.right = tInsert(node.right, key)
	} else {
		return node
	}
	return rebalance(node)
}

func tDel(node *AVLNode, key string) *AVLNode {
	if node == nil {
		return node
	}
	if key < node.key {
		node.left = tDel(node.left, key)
	} else if key > node.key {
		node.right = tDel(node.right, key)
	} else {
		if node.left == nil || node.right == nil {
			var temp *AVLNode
			if node.left != nil {
				temp = node.left
			} else {
				temp = node.right
			}
			if temp == nil {
				node = nil
			} else {
				*node = *temp
			}
		} else {
			temp := minValueNode(node.right)
			node.key = temp.key
			node.right = tDel(node.right, temp.key)
		}
	}
	if node == nil {
		return node
	}
	return rebalance(node)
}

func tGet(node *AVLNode, key string) *AVLNode {
	if node == nil || node.key == key {
		return node
	}
	if key < node.key {
		return tGet(node.left, key)
	}
	return tGet(node.right, key)
}

func rebalance(node *AVLNode) *AVLNode {
	node.height = 1 + max(height(node.left), height(node.right))
	balance := getBalance(node)

	if balance > 1 && getBalance(node.left) >= 0 {
		return rightRotate(node)
	}
	if balance > 1 && getBalance(node.left) < 0 {
		node.left = leftRotate(node.left)
		return rightRotate(node)
	}
	if balance < -1 && getBalance(node.right) <= 0 {
		return leftRotate(node)
	}
	if balance < -1 && getBalance(node.right) > 0 {
		node.right = rightRotate(node.right)
		return leftRotate(node)
	}
	return node
}

func leftRotate(z *AVLNode) *AVLNode {
	y := z.right
	t2 := y.left
	y.left = z
	z.right = t2
	z.height = 1 + max(height(z.left), height(z.right))
	y.height = 1 + max(height(y.left), height(y.right))
	return y
}

func rightRotate(z *AVLNode) *AVLNode {
	y := z.left
	t3 := y.right
	y.right = z
	z.left = t3
	z.height = 1 + max(height(z.left), height(z.right))
	y.height = 1 + max(height(y.left), height(y.right))
	return y
}

func getBalance(n *AVLNode) int {
	if n == nil {
		return 0
	}
	return height(n.left) - height(n.right)
}

func minValueNode(node *AVLNode) *AVLNode {
	current := node
	for current.left != nil {
		current = current.left
	}
	return current
}

func tPrint(node *AVLNode, indent string, last bool) {
	if node != nil {
		fmt.Print(indent)
		if last {
			fmt.Print("R----")
			indent += "   "
		} else {
			fmt.Print("L----")
			indent += "|  "
		}
		fmt.Printf(`"%s" (h:%d, b:%d)\n`, node.key, node.height, getBalance(node))
		tPrint(node.left, indent, false)
		tPrint(node.right, indent, true)
	}
}
